language/stdlib/modules/libra_system_2.move

```diff
@@ -0,0 +1,264 @@
+        _  = Vector::swap_remove(
+            &mut validator_set_ref.validators,
+            to_remove_index
+        );
+        emit_reconfiguration_event();
+    }
+    
+    // Copies validator's config to ValidatorSet/
+    // Callable by the LibraAssociation address only.
+    // Returns true if some portion of the config was changed and false otherwise
+    // Aborts if there is no change to the config.
+    public fun update_validator_info(addr: &address) acquires ValidatorSet {
+        Transaction::assert(Transaction::sender() == 0xA550C18, 1);
+        let validator_set_ref = borrow_global_mut<ValidatorSet>(0x1D8);
+
+        Transaction::assert(is_validator_(addr, &validator_set_ref.validators), 19);
+        let i = get_validator_index_(&validator_set_ref.validators, addr);
+
+        Transaction::assert(update_ith_validator_info_(&mut validator_set_ref.validators, i), 42);
+        emit_reconfiguration_event();
+    }
+
+    // This function can be called by anyone in posession of a valid ValidatorConfigUpdateToken,
+    // only validators or validators' delegated addresses can get this token as an output of
+    // rotate_consensus_key function.
+    // Returns true if the config was updated and false if config stays unchanged
+    public fun force_update_validator_info(update_token: ValidatorConfig2::ValidatorConfigUpdateToken) acquires ValidatorSet {
+        let validator_vec_ref = borrow_global_mut<ValidatorSet>(0x1D8);
+        // check the token is for one of the current validators
+        let validator_address = ValidatorConfig2::get_token_address(&update_token);
+        Transaction::assert(is_validator_(validator_address, &validator_vec_ref.validators), 19);
+
+        // check that the sequence number is correct to make sure the token can only be used once
+        Transaction::assert(ValidatorConfig2::get_token_sequence_number(&update_token) + 1 ==
+                            ValidatorConfig2::get_sequence_number(&ValidatorConfig2::get_config(*validator_address)), 42);
+        
+        let i = get_validator_index_(&validator_vec_ref.validators, validator_address);
+        let validator_info_ref = Vector::borrow(&validator_vec_ref.validators, i);
+        let current_block_time = LibraTimestamp::now_microseconds();
+        // Destroy token
+        ValidatorConfig2::destroy_update_token(update_token);
+
+        // If less than 24hr has passed since the last forced update, abort.
+        // Note that validators can always ask LibraAssocition offchain to invoke update.
+        Transaction::assert(current_block_time - validator_info_ref.last_force_update_time >= 86400000000, 11);
+
+        // Assert if there are no changes to the config
+        Transaction::assert(update_ith_validator_info_(&mut validator_vec_ref.validators, i), 42);
+        let validator_info = Vector::borrow_mut(&mut validator_vec_ref.validators, i);
+        *&mut validator_info.last_force_update_time = current_block_time;
```

Comment:
why does this update time but update_validator_info does not?

Reply:
Great question! Here my logic:

This force_update_* function can be used by a validator who just rotated its key and wants to immediately update the key in the validator_set, that's **the only function** that the validator can call in this module (I should've named it better). Other functions are only callable by the Association, the ValidatorSet address or the VM. And we want to rate-limit each validator in calls to this function.

So my logic was that Association's updates should not reset this timer. This way the validator can call a force-update any time (basically) even right after the Association did an update itself - to recover from a compromise, meaning only in some extreme circumstances. But we still want to rate limit - in case the validator experience a global attack and it's the attacker who in fact calls the force_update_*.